File: core/src/main/java/org/serio/core/showscrawler/crawler/step/FetchStep.java
package org.serio.core.showscrawler.crawler.step;

import java.util.Objects;

/**
 * Treats array of input strings from the previous steps as URLs and fetches their contents. Returns a corresponding
 * array of raw string response bodies.
 */
public class FetchStep implements CrawlerStep {
    public static final String TYPE = "fetch";

    /**
     * {@inheritDoc}
     */
    @Override
    public String getType() {
        return TYPE;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        return o != null && getClass() == o.getClass();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int hashCode() {
        return Objects.hashCode(TYPE);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString() {
        return "FetchStep{}";
    }
}

File: core/src/tv-show-crawler-runtime/executor/FetchStep.cpp
#include "tv-show-crawler-runtime/executor/FetchStep.h"

namespace serio::core {

std::vector<std::string> executeFetchStep(const CrawlerStep &step, const std::vector<std::string> &previousStepResults,
                                          HttpClient &httpClient) {
    return httpClient.fetchContentFromLinks(previousStepResults).get();
}

}